### PR TITLE
support prop:procedure

### PR DESCRIPTION
**Pull request by capfredf**
see the commit messages

### CODE REVIEW

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 1]
```diff
@@ -11,6 +11,7 @@
          "../private/syntax-properties.rkt"
          "../types/base-abbrev.rkt"
          "../types/abbrev.rkt"
+         (only-in "../types/numeric-tower.rkt" -Index)
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:14]: not used

File: typed-racket-lib/typed-racket/rep/rep-utils.rkt [hunk 2]
```diff
@@ -216,9 +216,10 @@
     [(_ rep ins fld ... update-fn)
      (with-syntax ([(new-val ...) (generate-temporaries #'(fld ...))]
                    [(old-val ...) (generate-temporaries #'(fld ...))])
-       #'(begin (match-define (struct* rep ([fld old-val] ...)) ins)
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/rep/rep-utils.rkt:219]:
```
(let ()
 ....)
```

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 3]
```diff
@@ -86,8 +86,35 @@
...
+                           null)))
+
+;; make a struct and its substructs executable
+(define/cond-contract (mk-substructs-procedural! base-tname base-sty new-ty)
+  (c:-> identifier? Struct? Fun? void?)
+  (define substuct-tnames (free-id-table-ref inheritance-tbl base-tname null))
+  (for ([stn (in-list substuct-tnames)])
+    (define sty
+      (match (lookup-type-name stn)
+        [(Intersection: (list-no-order (? Struct? sty) (? Type?) ...) _) sty]
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:114]: replace list-no-order to for/first

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 4]
```diff
@@ -86,8 +86,35 @@
...
+        [(? Struct? sty) sty]))
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:115]: one more case for more than one structs

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 5]
```diff
@@ -152,14 +242,15 @@
...
-                            ((or (Poly? parent) (Mu? parent)
-                                 (if prefab? (Prefab? parent) (Struct? parent)))
-                             parent)
-                            (else
+                          (match parent
+                            [(? Name?) (loop (resolve-name parent))]
+                            [(or (? Poly?) (? Mu?) (? Struct?)) parent]
+                            [(? Prefab?) #:when prefab?
+                                         parent]
+                            [(Intersection: (list-no-order (? Struct? parent) (? Type?) ...) _) parent]
```
[the remark this answers is not in the document]
Comment (capfredf) [reply typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:250]: I took another look at this line. This actually is a residue of my previous implementation, where a struct type refers to an intersection type of the function type and its actual structure type.  Under the current implementation, we will never run into this case.

TL;DR we are good without this case.

File: typed-racket-doc/typed-racket/scribblings/reference/special-forms.scrbl [hunk 6]
```diff
@@ -814,5 +804,45 @@ Performs type checking of forms entered at the read-eval-print loop.  The
...
+  (struct-field-index b))
+(animal 2 add1)
+(struct plant ([a : Number])
+  #:property prop:procedure
+  (lambda ([me : plant] [a1 : String]) : Number
+    (+ (plant-a me) (string-length a1))))
+(plant 31)
+]
+
+When a structure's base structure is procedural, so is the structure.
```
Comment (samth) [inline typed-racket-doc/typed-racket/scribblings/reference/special-forms.scrbl:828]: I think "is procedural" is a bad phrase -- just spell out what we mean.

File: typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt [hunk 7]
```diff
@@ -628,6 +628,8 @@
   (syntax-parse stx
     [(pmb . forms) (begin0 (type-check #'forms) (do-time "finished type checking"))]))
 
+
+(define toplevel-struct-making-tbl (make-free-id-table))
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt:632]: What is this table?

File: typed-racket-test/unit-tests/interactive-tests.rkt [hunk 8]
```diff
@@ -198,6 +198,19 @@
     (test-form #rx"not in the given function's range"
       (:query-type/result syntax-local-expand-expression Boolean))
 
+    (test-form-exn #rx"expected: \\(-> john-doe Output-Port \\(U Boolean One Zero\\) AnyValues\\)"
```
Comment (samth) [inline typed-racket-test/unit-tests/interactive-tests.rkt:201]: `regexp-quote` might be nicer here

File: typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt [hunk 9]
```diff
@@ -639,19 +641,38 @@
...
+       (define parsed (parse-typed-struct form))
+       (register-parsed-struct-sty! parsed)
+       (refine-struct-variance! (list parsed))
+       (register-parsed-struct-bindings! parsed))
+     (define all-forms (cond
+                         [(typed-struct? form)
+                          ;; after a struct type is registered, check the pending forms in order in which they are parsed.
+                          (define st-tname (name-of-struct form))
+                          (begin0 (append
+                                   (reverse (free-id-table-ref toplevel-struct-making-tbl st-tname))
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt:667]: You can save a list traversal by doing `cons` before `reverse`

File: typed-racket-doc/typed-racket/scribblings/reference/special-forms.scrbl (hunk 6 above)
Comment (capfredf) [reply typed-racket-doc/typed-racket/scribblings/reference/special-forms.scrbl:828]: use "has a prop:procedure" or something

**Review by samth — APPROVED** [review]
